pennylane/qnode.py

```diff
@@ -299,29 +302,36 @@ def construct(self, args, **kwargs):
         # check the validity of the circuit
 
         # quantum circuit function return validation
-        if isinstance(res, pennylane.operation.Expectation):
+        if isinstance(res, pennylane.operation.Observable):
             self.output_type = float
             self.output_dim = 1
             res = (res,)
-        elif isinstance(res, Sequence) and res and all(isinstance(x, pennylane.operation.Expectation) for x in res):
-            # for multiple expectation values, any valid Python sequence of expectation values (i.e., lists, tuples, etc) are supported in the QNode return statement.
+        elif isinstance(res, Sequence) and res and all(isinstance(x, pennylane.operation.Observable) for x in res):
+            # for multiple observables values, any valid Python sequence of observables
+            # (i.e., lists, tuples, etc) are supported in the QNode return statement.
             self.output_dim = len(res)
             self.output_type = np.asarray
             res = tuple(res)
         else:
-            raise QuantumFunctionError("A quantum function must return either a single expectation "
-                                       "value or a nonempty sequence of expectation values.")
+            raise QuantumFunctionError("A quantum function must return either a single measured observable "
+                                       "or a nonempty sequence of measured observables.")
+
+        # check that all returned observables have a return_type specified
+        for x in res:
+            if x.return_type is None:
+                raise QuantumFunctionError("Observable '{}' does not have the measurement "
+                                           "type specified.")
 
         # check that all ev's are returned, in the correct order
         if res != tuple(self.ev):
-            raise QuantumFunctionError("All measured expectation values must be returned in the "
+            raise QuantumFunctionError("All measured observables must be returned in the "
                                        "order they are measured.")
 
-        self.ev = res  #: tuple[Expectation]: returned expectation values
+        self.ev = res  #: tuple[Observable]: returned observables
         self.ops = self.queue + list(self.ev)  #: list[Operation]: combined list of circuit operations
 
         # classify the circuit contents
-        temp = [isinstance(op, pennylane.operation.CV) for op in self.ops]
+        temp = [isinstance(op, pennylane.operation.CV) for op in self.ops if not isinstance(op, pennylane.ops.Identity)]
```

Comment:
how come you added a conditional on the identity here?

Reply:
When we first coded up PennyLane, we didn't allow for operations that were both DV _and_ CV 😆 

Because of inheritence, an `Identity(CVObservable, Observable)` instance will be such that `isinstance(op, pennylane.operation.CV) == True`. So using it in a qubit circuit would cause the `QuantumFunctionError` below to be triggered.

Should make this a TODO to better explore how to provide operations with the same name in both qubit/CV circuits (although I expect `Identity` will be the only case this occurs)